src/bun.js/api/glob.zig

```diff
@@ -0,0 +1,381 @@
+        return JSC.JSArray.from(globalThis, &[_]JSC.JSValue{});
+    }
+
+    return BunString.toJSArray(globalThis, globWalk.matchedPaths.items[0..]);
+}
+
+fn makeGlobWalker(
+    this: *Glob,
+    globalThis: *JSGlobalObject,
+    arguments: *ArgumentsSlice,
+    comptime fnName: []const u8,
+    alloc: Allocator,
+    arena_: Arena,
+) ?*GlobWalker {
+    var arena = arena_;
+    const matchOpts = MatchOpts.fromJS(globalThis, arguments, fnName, &arena);
+    if (matchOpts != null and matchOpts.?.cwd != null) {
+        var globWalker = alloc.create(GlobWalker) catch {
+            globalThis.throw("Out of memory", .{});
+            return null;
+        };
+
+        globWalker.* = .{};
+        globWalker.initWithCwd(arena, this.pattern, matchOpts.?.cwd.?, matchOpts.?.dot, matchOpts.?.absolute) catch {
+            globalThis.throw("Out of memory", .{});
+            return null;
+        };
+        return globWalker;
+    }
+    var globWalker = alloc.create(GlobWalker) catch {
+        globalThis.throw("Out of memory", .{});
+        return null;
+    };
+
+    globWalker.* = .{};
+    switch (globWalker.init(arena, this.pattern, false, false) catch {
+        globalThis.throw("Out of memory", .{});
+        return null;
+    }) {
+        .err => |err| {
+            globalThis.throwValue(err.toJSC(globalThis));
+            return null;
+        },
+        else => {},
+    }
+
+    return globWalker;
+}
+
+pub fn constructor(
+    globalThis: *JSC.JSGlobalObject,
+    callframe: *JSC.CallFrame,
+) callconv(.C) ?*Glob {
+    const alloc = getAllocator(globalThis);
+
+    const arguments_ = callframe.arguments(1);
+    var arguments = JSC.Node.ArgumentsSlice.init(globalThis.bunVM(), arguments_.slice());
+    defer arguments.deinit();
+    const pat_arg = arguments.nextEat() orelse {
+        globalThis.throw("Glob.constructor: expected 1 arguments, got 0", .{});
+        return null;
+    };
+
+    if (!pat_arg.isString()) {
+        globalThis.throw("Glob.constructor: first argument is not a string", .{});
+        return null;
+    }
+
+    var pat_str: []u8 = pat_arg.getZigString(globalThis).toOwnedSlice(globalThis.bunVM().allocator) catch @panic("OOM");
+
+    const all_ascii = isAllAscii(pat_str);
+
+    var glob = alloc.create(Glob) catch @panic("OOM");
+    glob.* = .{ .pattern = pat_str, .is_ascii = all_ascii };
+
+    if (!all_ascii) {
+        var codepoints = std.ArrayList(u32).initCapacity(alloc, glob.pattern.len * 2) catch {
+            globalThis.throwOutOfMemory();
+            return null;
+        };
+        errdefer codepoints.deinit();
+
+        convertUtf8(&codepoints, glob.pattern) catch {
+            globalThis.throwOutOfMemory();
+            return null;
+        };
+
+        glob.pattern_codepoints = codepoints;
+    }
+
+    return glob;
+}
+
+pub fn finalize(
+    this: *Glob,
+) callconv(.C) void {
+    const alloc = JSC.VirtualMachine.get().allocator;
+    alloc.free(this.pattern);
```

Comment:
What keeps the `pattern` alive if the `Glob` object is freed before the `Promise` for an async task resolves successfully? You might need to use `hasPendingActivity` and make it return `true` when promises have not been fulfilled yet.

Reply:
Something like this should reproduce it:

```zig
import {Glob} from 'bun';
const promise = (async function() {
  const glob = new Glob("**/*.zig")
  return glob.match("/");
});
Bun.gc(true);
await promise;
```